keps/sig-apps/4017-pod-index-label/README.md

```diff
@@ -0,0 +1,898 @@
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+The StatefulSet controller will only need a minor update to the [newStatefulSetPod](https://github.com/kubernetes/kubernetes/blob/fb5e9ef3b2a6f2136d54868187431c345e59f55f/pkg/controller/statefulset/stateful_set_utils.go#L458) function, to set the pod ordinal as the label `statefulset.kubernetes.io/pod-ordinal`. This call is downstream from the [newVersionedStatefulSetPod](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L416) call, which generates
+the StatefulSet pods before creating them as necessary in [CreateStatefulPod](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L433).
+
+Similarly, the Job controller would need to add the completion index as a label [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L1481)
+where it adds the corresponding annotation.
+
+Next, we must determine how enabled this new feature will interact with existing workloads. There are a couple of options:
+
+1. Only inject the label on *newly created pods*, so an existing StatefulSet/Indexed Jobs may include pods with the label and some without it. 
+This means for the user to utilize the label via the downard API, or to use the label for pod selection, they will need to recreate 
+the StatefulSet so the label is present on all pods.
+
+2. Inject the label only on pods for *newly created StatefulSets/Indexed Jobs*. We can track this by annotating StatefulSets/Indexed Jobs. This way,
+for a given StatefulSet/IndexedJob, either none of the pods have this label, or all of them do, which will provide a more consistent
+user experience. However, in the case of a cluster downgrade to a version without this feature, new pods would start getting
+created without this label again.
+
+Both options will not be disruptive to existing workloads, however, I lean towards option 2 so outside of the cluster downgrade edge case,
+there will be consistency within a single StatefulSet/Indexed Job and therefore a more predictable user experience.
```

Comment:
adding an annotation is hacky though and it will carry no meaning in the future, but we may be stuck with it because it kinda become part of the api. I would go with option 1 for simplicity.

Reply:
Discussed this offline and updated the recommendation accordingly